MatrixSDK/Data/MXEventTimeline.m

```diff
@@ -167,73 +221,77 @@ - (MXHTTPOperation *)paginate:(NSUInteger)numItems direction:(MXTimelineDirectio
-                                                                     // Check whether the pagination end is reached
-                                                                     MXError *mxError = [[MXError alloc] initWithNSError:error];
-                                                                     if (mxError && [mxError.error isEqualToString:kMXErrorStringInvalidToken])
-                                                                     {
-                                                                         // We run out of items
-                                                                         [store storeHasReachedHomeServerPaginationEndForRoom:_state.roomId andValue:YES];
-                                                                         
-                                                                         NSLog(@"[MXEventTimeline] paginate: pagination end has been reached");
-                                                                         
-                                                                         // Ignore the error
-                                                                         complete();
-                                                                         return;
-                                                                     }
-                                                                     
-                                                                     NSLog(@"[MXEventTimeline] paginate error: %@", error);
-                                                                     failure(error);
-                                                                 }];
-            
-            if (messagesFromStoreCount)
-            {
-                // Disable retry to let the caller handle messages from store without delay.
-                // The caller will trigger a new pagination if need.
-                operation.maxNumberOfTries = 1;
-            }
-        }
-        else
+        if (0 == numItems || YES == [store hasReachedHomeServerPaginationEndForRoom:_state.roomId])
         {
             // Nothing more to do
             complete();
-            
+
             NSLog(@"[MXEventTimeline] paginate: is done");
+            return nil;
         }
     }
-    else if (!_isLiveTimeline)
+
+    // Not enough messages: make a pagination request to the home server
+    // from last known token
+    NSString *paginationToken;
+
+    if (direction == MXTimelineDirectionBackwards)
     {
-        NSAssert(NO, @"TODO: paginate forwards is not yet supported");
+        paginationToken = [store paginationTokenOfRoom:_state.roomId];
+        if (nil == paginationToken)
+        {
+            paginationToken = @"END";
+        }
     }
     else
     {
-        NSAssert(NO, @"Cannot paginate forwards on a live timeline");
+        paginationToken = forwardsPaginationToken;
```

Comment:
Why not check hasReachedHomeServerForwardsPaginationEnd flag before triggering forward pagination?